Add UI for bypassing SSL handshake failures

jdm opened the pull request:
There are several parts to these changes:
1. resurrecting the network error classification code to distinguish between SSL failures and other network errors
1. adding an SSL verification callback to support verifying certs against a list that can change at runtime, rather than just at program initialization
1. exposing a privileged chrome://allowcert URI which accepts the PEM cert contents along with a secret token
1. extracting the PEM cert contents out of the network layer when a handshake failure occurs, and getting them into the HTML that is parsed when an SSL failure occurs
1. adding a button in the handshake failure page that performs an XHR to chrome://allowcert with knowledge of the secret token and the PEM cert contents, before reloading the original URL that failed

The presence of the secret token means that while the chrome://allowcert URL is currently visible to web content, they cannot make use of it to inject arbitrary certs into the verification process.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #26683
- [x] These changes do not require tests because the UI requires user activation and can't clearly be automated

jdm committed:
net: Treat SSL handshake errors differently from other hyper errors.

jdm committed:
net: Add an SSL verification callback to support checking a dynamic list of certs.

jdm committed:
net: Pass certs that fail the SSL handshake out of the network layer.

jdm committed:
net: Allow SSL websockets to use dynamic list of certs as well.

jdm committed:
net: Add option to temporarily accept certs that failed the handshake.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/servoparser/mod.rs
 * @kichjang: components/net/fetch/methods.rs, components/net/resource_thread.rs, components/net/http_loader.rs, components/net/websocket_loader.rs, components/script/dom/servoparser/mod.rs and 3 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify net and script code, but no tests are modified. Please consider adding a test!

jdm commented:
<img width="1136" alt="Screen Shot 2020-05-29 at 1 35 56 PM" src="https://user-images.githubusercontent.com/27658/83289271-d2234980-a1b2-11ea-9c28-12d66cb8fb3c.png">

Here's an example of loading the default Hubs local dev server with the new UI. Pressing the button reloads the page without a handshake error.

asajeffrey commented on components/net/connector.rs:
```diff
@@ -30,6 +35,38 @@ const SIGNATURE_ALGORITHMS: &'static str = concat!(
+    fn store(&self, host: String, cert_bytes: Vec<u8>) {
+        let mut certs = self.certs.lock().unwrap();
+        let entry = certs.entry(host).or_insert((cert_bytes, 0));
+        entry.1 += 1;
```
This looks a lot like a poor dev's weak Rc. Is it?

jdm replied:
It does have the property of removing the hashmap entry when the last user goes away, so it's slightly more complex than Rc/Arc.

Manishearth replied:
Yeah I could go in either direction here

asajeffrey commented on resources/badcert.html:
```diff
@@ -3,6 +3,24 @@
-    <p>${reason}</p>
+  <p>${reason}</p>
+  <pre id="bytes">${bytes}</pre>
+  <button id="allow">Allow certificate temporarily</button>
```
Should there also be a "decline" button?

jdm replied:
Perhaps. What should it do, however?

jdm replied:
Firefox does have the Get Me Out Of Here button which I think does the same thing, so I'll add that.

asajeffrey commented on resources/badcert.html:
```diff
@@ -3,6 +3,24 @@
+    if (bytes.length) {
+        button.onclick = function() {
+            let xhr = new XMLHttpRequest();
+            xhr.open('GET', 'chrome:allowcert?secret=${secret}&bytes=' + btoa(bytes));
```
I'm a bit concerned that we're using a GET request for this, side-effecting GET that adds a security wormhole scares me!

jdm replied:
What is the security wormhole, exactly?

Manishearth replied:
Yeah IMO this should be a POST. Secrets in url params is bad

asajeffrey reviewed (commented):
This all looks like a sensible approach, apart from using a side-effecting GET as the way of bypassing cert checks.

jdm commented:
Note to self: unit tests don't build now.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26678) made this pull request unmergeable. Please resolve the merge conflicts.

Manishearth reviewed (approved):
r=me with minor changes, and we should also be passng unit tests

jdm committed:
net: Use a POST request for allowing certs temporarily.

jdm committed:
net: Add unit test for accepting a self-signed cert.

jdm commented:
Fixed the unit tests, added on for accepting self-signed certs, and added a button to go back in history on the cert error page.
@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit 33c5f54 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 33c5f54847b7b5fa067c0c01c557468a8992d3c9 -->

jdm commented:
@bors-servo r=Manishearth,asajeffrey

bors-servo commented:
:pushpin: Commit ae447e9 has been approved by `Manishearth,asajeffrey`

<!-- @bors-servo r=Manishearth,asajeffrey ae447e93c5ed5642f6ae2b3d0d35cbb161ec343f -->

bors-servo commented:
:hourglass: Testing commit ae447e93c5ed5642f6ae2b3d0d35cbb161ec343f with merge ab7a72dadd95e2b88ff6fb203b9476cc787e6190...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ebB-h_VqTAWx_2679SqinA)

jdm commented:
@bors-servo r=Manishearth,asajeffrey

bors-servo commented:
:pushpin: Commit d8fb5a3 has been approved by `Manishearth,asajeffrey`

<!-- @bors-servo r=Manishearth,asajeffrey d8fb5a34669e94f397ed80f52135d07847c92449 -->

bors-servo commented:
:hourglass: Testing commit d8fb5a34669e94f397ed80f52135d07847c92449 with merge 5c24d407037dba27721e8baa5a37bcf86a7d8462...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/KsZiW3yxSPihOpA92rv4YA)

jdm commented:
@bors-servo r=Manishearth,asajeffrey

bors-servo commented:
:pushpin: Commit dce675d has been approved by `Manishearth,asajeffrey`

<!-- @bors-servo r=Manishearth,asajeffrey dce675dae3a161fa1aa99d07a15a23d55049d8c3 -->

bors-servo commented:
:hourglass: Testing commit dce675dae3a161fa1aa99d07a15a23d55049d8c3 with merge efab9644220b901d95c15d6fb8bc37ec60f6cdda...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ek7-F8KTQWegPWtwmf2-eQ)

jdm commented:
@bors-servo r=Manishearth,asajeffrey

bors-servo commented:
:pushpin: Commit c8692d8 has been approved by `Manishearth,asajeffrey`

<!-- @bors-servo r=Manishearth,asajeffrey c8692d83ab46899dc3ed3f58388164df50b485b9 -->

bors-servo commented:
:hourglass: Testing commit c8692d83ab46899dc3ed3f58388164df50b485b9 with merge 389246300fac8204e7e124f7b472ea637f212e99...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/SCFjMT9wS0aDkjcQMcdvcw)

jdm commented:
@bors-servo retry
#25513

bors-servo commented:
:hourglass: Testing commit c8692d83ab46899dc3ed3f58388164df50b485b9 with merge 7fa5bf4da24f05e005600854ce9c6ad7815b4f74...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/F16i4vukTreUcNKT_1Yh4A)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit c8692d83ab46899dc3ed3f58388164df50b485b9 with merge 0b0ea17dca72d867b56ddf518240e25f30d93f3e...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/J3AZp0EERVi3MWhiezXCFw)
Approved by: Manishearth,asajeffrey
Pushing 0b0ea17dca72d867b56ddf518240e25f30d93f3e to master...
